async-usercalls/src/hacks/unsafe_typecasts.rs

```diff
@@ -0,0 +1,95 @@
+//! The incredibly unsafe code in this module allows us to create
+//! `std::net::TcpStream` and `std::net::TcpListener` types from their raw
+//! components in SGX.
+//!
+//! This is obviously very unsafe and not maintainable and is only intended as
+//! an iterim solution until we add similar functionality as extension traits
+//! in `std::os::fortanix_sgx`.
```

Comment:
Sure. How do you feel about adding the following trait instead of changing `FromRawFd`?

```rust
pub trait FromRawComponents {
    type Components;

    unsafe fn from_raw_components(c: Self::Components) -> Self;
}

impl FromRawComponents for TcpListener {
    type Components = (RawFd, Option<String>);

    unsafe fn from_raw_components(c: Self::Components) -> Self {
        // ...
    }
}

impl FromRawComponents for TcpStream {
    type Components = (RawFd, Option<String>, Option<String>);

    unsafe fn from_raw_components(c: Self::Components) -> Self {
        // ...
    }
}
```

Reply:
As discussed over the phone, add an associated `type Metadata` to `FromRawFd`.